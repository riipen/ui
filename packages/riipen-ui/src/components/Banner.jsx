import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class Banner extends React.Component {
  static propTypes = {
    /**
     * Content to render inside the Banner.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The height to render the Banner at.
     */
    height: PropTypes.string,

    /**
     * The image source to display in the Banner.
     */
    src: PropTypes.string,

    /**
     * A tint color to apply over the Banner's image.
     */
    tint: PropTypes.oneOf([
      "primary",
      "secondary",
      "tertiary",
      "positive",
      "negative"
    ])
  };

  static contextType = ThemeContext;

  render() {
    const { children, classes, height, src, tint } = this.props;

    const theme = this.context;

    const className = clsx("banner", tint, classes);

    return (
      <React.Fragment>
        <div className={className}>{children}</div>
        <style jsx>{`
          .banner {
            background-image: url(${src});
            background-position: center;
            background-size: cover;
            display: flex;
            height: ${height};
            position: relative;
          }
          .banner::after {
            content: "";
            top: 0;
            left: 0;
            height: 100%;
            width: 100%;
            opacity: 0.9;
            position: absolute;
            z-index: ${theme.zIndex.lowest};
          }
          .banner > :global(*) {
            z-index: ${theme.zIndex.low};
          }

          .primary::after {
            background-color: ${theme.palette.primary.main};
          }
          .secondary::after {
            background-color: ${theme.palette.secondary.main};
          }
          .tertiary::after {
            background-color: ${theme.palette.tertiary.main};
          }
          .positive::after {
            background-color: ${theme.palette.positive.main};
          }
          .negative::after {
            background-color: ${theme.palette.negative.main};
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Banner);
