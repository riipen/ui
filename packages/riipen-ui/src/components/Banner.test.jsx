import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Banner from "./Banner.jsx";

describe("<Banner>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Banner />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Banner />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("renders given children", () => {
      const child = <h5>hi hi</h5>;

      const wrapper = mount(<Banner children={child} />);

      expect(
        wrapper
          .find("Banner")
          .childAt(0)
          .contains(child)
      ).toEqual(true);
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the div inside of Banner", () => {
      const classes = ["classOne", "classTwo"];

      const wrapper = mount(<Banner classes={classes} />);

      expect(
        wrapper
          .find("Banner")
          .childAt(0)
          .find("div")
          .hasClass(classes[0])
      ).toEqual(true);
      expect(
        wrapper
          .find("Banner")
          .childAt(0)
          .find("div")
          .hasClass(classes[1])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-banner"];

      const wrapper = mount(<Banner />);

      expect(
        wrapper
          .find("Banner")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Banner classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("height prop", () => {
    it("sets the div inside Banner to a given height", () => {
      const height = "50px";

      const wrapper = mount(<Banner height={height} />);

      expect(
        wrapper
          .find("Banner")
          .childAt(0)
          .props().style.height
      ).toEqual(height);
    });

    it("gives an error when height is not given as a string", () => {
      const height = 50;
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Banner height={height} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("src prop", () => {
    it("sets background image with given src value", () => {
      const src = "/static/images/banners/folders.jpg";

      const wrapper = mount(<Banner src={src} />);

      expect(
        wrapper
          .find("Banner")
          .childAt(0)
          .props().style.backgroundImage
      ).toContain(src);
    });
  });

  describe("tint prop", () => {
    it("applies tint class name given valid tint", () => {
      const tint = "tertiary";

      const wrapper = mount(<Banner tint={tint} />);

      expect(
        wrapper
          .find("Banner")
          .childAt(0)
          .find("div")
          .hasClass(tint)
      ).toEqual(true);
    });

    it("gives an error when given an invalid tint", () => {
      const tint = "hi hi hi";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Banner tint={tint} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });
});
