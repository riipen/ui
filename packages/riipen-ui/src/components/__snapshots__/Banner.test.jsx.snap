// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner> renders correct snapshot 1`] = `
<withClasses(Banner)>
  <Banner
    classes={
      Array [
        "riipen",
        "riipen-banner",
      ]
    }
  >
    <div
      className="jsx-1351289181 banner riipen riipen-banner"
      style={
        Object {
          "backgroundImage": "url(undefined)",
          "height": undefined,
        }
      }
    />
    <JSXStyle
      dynamic={
        Array [
          1,
          2,
          "#1f71ae",
          "#f58632",
          "#3caabb",
          "#4cc88d",
          "#e95353",
        ]
      }
      id="3877515695"
    />
  </Banner>
</withClasses(Banner)>
`;
